### PR TITLE
Long running operations

### PR DESCRIPTION
**Pull request by SeanFeldman**
Connects to https://github.com/Particular/V6Launch/issues/81
Connects to https://github.com/Particular/docs.particular.net/issues/1447
Connects to https://github.com/Particular/docs.particular.net/issues/1533

### CODE REVIEW

File: samples/azure/azure-service-bus-long-running/ASB_7/Endpoint1/app.config [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+<add name="ws2007HttpRelayBinding" type="Microsoft.ServiceBus.Configuration.WS2007HttpRelayBindingCollectionElement, Microsoft.ServiceBus, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+<add name="netTcpRelayBinding" type="Microsoft.ServiceBus.Configuration.NetTcpRelayBindingCollectionElement, Microsoft.ServiceBus, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+<add name="netOnewayRelayBinding" type="Microsoft.ServiceBus.Configuration.NetOnewayRelayBindingCollectionElement, Microsoft.ServiceBus, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+<add name="netEventRelayBinding" type="Microsoft.ServiceBus.Configuration.NetEventRelayBindingCollectionElement, Microsoft.ServiceBus, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+<add name="netMessagingBinding" type="Microsoft.ServiceBus.Messaging.Configuration.NetMessagingBindingCollectionElement, Microsoft.ServiceBus, Culture=neutral, PublicKeyToken=31bf3856ad364e35" />
+</bindingExtensions>
+</extensions>
+</system.serviceModel>
```
Comment (yvesgoeleven) [inline samples/azure/azure-service-bus-long-running/ASB_7/Endpoint1/app.config]: not needed, can be removed

File: samples/azure/azure-service-bus-long-running/ASB_7/Endpoint2/LongProcessingRequestSaga.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        #endregion
+
+        log.Info($"Timeout is set to be executed at {timeoutToBeInvokedAt}.");
+
+
+        log.Info("Registering long running request with Processor.");
+
+        #region enqueue-request-for-processor
+
+        // Saga enqueues the request in storate table. Normally, if there's more than that, work should be done by a handler
```
Comment (yvesgoeleven) [inline samples/azure/azure-service-bus-long-running/ASB_7/Endpoint2/LongProcessingRequestSaga.cs]: How about: Saga enqueues the request to process in a storage table. This is the logical equivalant of adding a message to a queue, if there would be business specific work in here, that work should be done by sending a message to a handler instead, not in the saga.

File: samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                    // process
+                    var estimatedProcessingTime = TimeSpan.Parse(request.EstimatedProcessingTime);
+                    await Task.Delay(estimatedProcessingTime, token).ConfigureAwait(false);
+                    log.Info($"Request with ID {request.RequestId} processed.");
+
+                    request.Status = Status.Finished.ToString();
+                    request.FinishedAt = DateTime.UtcNow;
+                    await table.ExecuteAsync(TableOperation.Merge(request), token).ConfigureAwait(false);
+                    toProccess.TryDequeue(out request);
+                    await endpoint.Publish<LongProcessingFinished>(m => m.Id = request.RequestId).ConfigureAwait(false);
```
Comment (yvesgoeleven) [inline samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs]: shouldn't the publish happen before removing the work from the queue?

File: samples/azure/azure-service-bus-long-running/ASB_7/Shared/LongProcessingRequestReply.cs [hunk 4]
```diff
@@ -0,0 +1,7 @@
+using System;
+using NServiceBus;
+
+public class LongProcessingRequestReply : IMessage
```
Comment (yvesgoeleven) [inline samples/azure/azure-service-bus-long-running/ASB_7/Shared/LongProcessingRequestReply.cs]: I would name this LongProcessingReply

**Comment by yvesgoeleven** [comment]
Added a few comments, besides that looks good to me!

File: samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs (hunk 3 above)
Comment (SeanFeldman) [inline samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs]: Discussed with @yvesgoeleven , all good. Saga covers data in the storage table.

File: samples/azure/azure-service-bus-long-running/ASB_7/Endpoint1/Client.csproj [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    <Reference Include="Newtonsoft.Json, Version=9.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
+      <HintPath>..\..\..\..\..\packages\Newtonsoft.Json.9.0.1\lib\net45\Newtonsoft.Json.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="NServiceBus.Azure.Transports.WindowsAzureServiceBus, Version=7.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c, processorArchitecture=MSIL">
+      <HintPath>..\..\..\..\..\packages\NServiceBus.Azure.Transports.WindowsAzureServiceBus.7.0.0-rc0001\lib\net452\NServiceBus.Azure.Transports.WindowsAzureServiceBus.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="NServiceBus.Core">
+      <HintPath>..\..\..\..\..\packages\NServiceBus.6.0.0-rc0001\lib\net452\NServiceBus.Core.dll</HintPath>
```
Comment (SimonCropp) [inline samples/azure/azure-service-bus-long-running/ASB_7/Endpoint1/Client.csproj]: rc2 :)

Comment (SeanFeldman) [inline samples/azure/azure-service-bus-long-running/ASB_7/Endpoint1/Client.csproj]: One thing at a time...

**Comment by SeanFeldman** [comment]
@Particular/docs-maintainers @Particular/azure-service-bus-maintainers wip removed. Please review.

File: samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    static readonly ConcurrentQueue<RequestRecord> toProcess = new ConcurrentQueue<RequestRecord>();
+
+    public void Start(IEndpointInstance endpointInstance, CancellationToken token)
+    {
+        endpoint = endpointInstance;
+        var cloudTableClient = StorageHelper.GetTableClient();
+        table = cloudTableClient.GetTableReference(Constants.TableName);
+        table.CreateIfNotExists();
+
+        var task = new Task(() =>
```
Comment (SimonCropp) [inline samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs]: @danielmarbach what is the best way of doing the below? ie start two tasks and return immediately

Comment (SeanFeldman) [inline samples/azure/azure-service-bus-long-running/ASB_7/Processor/Processor.cs]: @SimonCropp that's exactly the idea. I need processor to watch changes and keep processing.

**Comment by SeanFeldman** [comment]
Build is green again.
@danielmarbach if you can suggest a better approach for the comment [here](https://github.com/Particular/docs.particular.net/pull/1831/#discussion_r78696226), please do.

**Pull request merged**